plugin/plugin.go

```diff
@@ -8,68 +8,65 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"io"
+	"strings"
 
 	"github.com/drone/drone-go/plugin/validator"
 	"github.com/open-policy-agent/opa/rego"
-	"gopkg.in/yaml.v2"
+	yaml "gopkg.in/yaml.v2"
 )
 
-// New returns a new validator plugin.
 func New(policy string) validator.Plugin {
 	return &plugin{
 		policyPath: policy,
 	}
 }
 
-type plugin struct {
-	policyPath string
-}
-
 func (p *plugin) Validate(ctx context.Context, req *validator.Request) error {
-	// Drone sends its raw config in the request body
-	type DroneConfig struct {
-		Kind  string `yaml:"kind" json:"kind"`
-		Type  string `yaml:"type" json:"type"`
-		Name  string `yaml:"name" json:"name"`
-		Steps []struct {
-			Name      string   `yaml:"name" json:"name"`
-			Image     string   `yaml:"image" json:"image"`
-			Commands  []string `yaml:"commands" json:"commands"`
-			DependsOn []string `yaml:"depends_on,omitempty" json:"depends_on,omitempty"`
-		} `yaml:"steps" json:"steps"`
-		Trigger struct {
-			Branch []string `yaml:"branch" json:"branch"`
-		} `yaml:"trigger" json:"trigger"`
-	}
 
-	var droneConfig DroneConfig
+	var documents []DroneConfig
 
-	err := yaml.Unmarshal([]byte(req.Config.Data), &droneConfig)
-	if err != nil {
-		return err
+	dec := yaml.NewDecoder(strings.NewReader(req.Config.Data))
+	for {
+		var document DroneConfig
+
+		err := dec.Decode(&document)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		documents = append(documents, document)
 	}
 
-	r := rego.New(
+	r, err := rego.New(
 		rego.Query("deny = data.drone.validation.deny; msg = data.drone.validation.out"),
-		rego.Load([]string{p.policyPath}, nil))
-
-	query, err := r.PrepareForEval(ctx)
+		rego.Load([]string{p.policyPath}, nil)).PrepareForEval(ctx)
 	if err != nil {
 		return err
 	}
 
-	rs, err := query.Eval(ctx, rego.EvalInput(droneConfig))
-	if err != nil {
-		return err
-	}
+	var message string
+	var values []string
 
-	if rs[0].Bindings["deny"] == true {
-		message := fmt.Sprintf("[validator] %v", rs[0].Bindings["msg"])
-	 	return errors.New(message)
+	for _, resource := range documents {
+		rs, err := r.Eval(ctx, rego.EvalInput(resource))
+		if err != nil {
+			return err
+		}
+		if rs[0].Bindings["deny"] == true {
+			s := strings.Split(fmt.Sprintf("%v", rs[0].Bindings["msg"]), ":")
+			if message == "" {
+				message = s[0]
+			}
```

Comment:
It is not entirely clear why `message` is only set for the very first error encountered.   Is it actually necessary? / Should all errors simply return as-is?

Reply:
It would be ideal, but the error box in the UI doesn't look good with multiple sentences. I added some more information about it on @colinhoglund 's comment below.